File: logic/src/main/java/com/tp/holdem/logic/HandOperations.java
package com.tp.holdem.logic;

import com.tp.holdem.logic.model.*;
import io.vavr.Tuple2;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import lombok.extern.slf4j.Slf4j;

import java.util.Comparator;
import java.util.function.Function;

@Slf4j
public class HandOperations {
	private static final Comparator<Card> CARD_COMPARATOR = Comparator.comparing(Card::getValue);

	private static HandRank findHandRank(Player player, PokerTable pokerTable) {
		final List<Card> cards = List.ofAll(player.getHand())
				.appendAll(pokerTable.getCardsOnTable())
				.sorted(CARD_COMPARATOR);

		final Hands playerHand = HandFinder.findHand(cards);
		final List<Card> bestCardsThatMakeHand = BestCardCombinationFinder.findBestCardsForHand(cards, playerHand);

		return new HandRank(playerHand, bestCardsThatMakeHand);
	}

	public static Tuple2<Player, HandRank> findWinner(List<Player> allPlayers, PokerTable pokerTable) {
		final Map<Player, HandRank> hands = allPlayers.filter(Player::playing)
				.toMap(Function.identity(), player -> findHandRank(player, pokerTable));

		log.debug(String.format("Players to hands map: %s", hands));

		return hands.values().maxBy(HandRankComparator.INSTANCE)
				.flatMap(handRank -> hands.find(tuple -> tuple._2.equals(handRank)))
				.getOrElseThrow(PlayerExceptions.PLAYER_NOT_FOUND);
	}
}

File: logic/src/main/kotlin/com/tp/holdem/logic/hands/HandOperations.kt
package com.tp.holdem.logic.hands

import com.tp.holdem.common.lazyLogger
import com.tp.holdem.logic.players.handRank
import com.tp.holdem.logic.players.notFolded
import com.tp.holdem.logic.players.playing
import com.tp.holdem.model.Player
import com.tp.holdem.model.PokerTable
import io.vavr.collection.List
import io.vavr.control.Either

internal object HandOperations {
    private val log by lazyLogger()

    fun findWinner(allPlayers: List<Player>, pokerTable: PokerTable): Either<Player, List<Player>> {
        val notFoldedPlayers = allPlayers.notFolded()

        if (notFoldedPlayers.size() == 1) {
            return Either.left<Player, List<Player>>(notFoldedPlayers.single())
                    .also { log.debug("Everyone folded except one player, he's the winner") }
        }

        val hands = allPlayers.playing()
                .toMap({ it }, { player -> player.handRank(pokerTable) })
                .also { log.debug("Players to hands map: $it") }

        val maxHandRank = hands.values()
                .maxBy(HandRankComparator)
                .single()

        val playersWithMaxHandRank = hands
                .filter { _, hand -> HandRankComparator.compare(hand, maxHandRank) == 0 }
                .map { tuple -> tuple._1 }
                .toList()

        return when (playersWithMaxHandRank.size()) {
            1 -> Either.left(playersWithMaxHandRank.head())
            else -> Either.right(playersWithMaxHandRank)
        }
    }
}
